File: R/example12_3.R
# Example 12.3 GMM Estimation of U.S. Consumption Function
# Using package AER, gmm

data<-read.table("http://web.pdx.edu/~crkl/ceR/data/usyc87.txt",header=T,nrows=66)
y<-ts(data$Y,start=1929,frequency=1)
c<-ts(data$C,start=1929,frequency=1)

# define lag operator L on one variable
L<-function(x,l) {c(rep(NA,l),x[1:(length(x)-l)])}

results.ols<-lm(c~L(c,1))
summary(results.ols)

library(AER)
# update/reinstall the following packages: rio,curl,data.table,magritter,hms,
# pkgconfig,crayon,stringi,zip,readxl,cellranger,abind,lmtest,sandwich,Formula
results.iv<-ivreg(c~L(c,1)|L(y,1))
summary(results.iv)

library(gmm)
# can npot run the following?
model1<-gmm(c~L(c,1),~L(y,1),results.ols$coefficient)  # default vcov="HAC"
summary(model1)
y1<-L(y,1)
c1<-L(c,1)
cc<-c[2:66]
c1<-c1[2:66]
y1<-y1[2:66]
yy<-y[2:66]
model1<-gmm(cc~c1,~y1,results.ols$coefficients)  # default vcov="HAC"
summary(model1)
model2<-gmm(cc~c1,~y1,results.ols$coefficients,type="iterative")
summary(model2)
model3<-gmm(cc~c1,~y1,results.ols$coefficients,type="cue")
summary(model3)

rm(list=ls())

File: python/example12_3.py
# Example 12.3 GMM Estimation of U.S. Consumption Function
# See also Example 11.2, using linearmodels package

import numpy as np
import pandas as pd
import statsmodels.api as sm
from linearmodels.iv import IV2SLS,IVGMM, IVGMMCUE

data = pd.read_csv("http://web.pdx.edu/~crkl/ceR/data/usyc87.txt",index_col='YEAR',sep='\s+',nrows=66)
y = data['Y']
c = data['C']
c1 = c.shift(1)
y1 = y.shift(1)
y2 = y.shift(2)
df = pd.DataFrame({"c":c,"c1":c1,"y":y,"y1":y1,"y2":y2})
df = df.dropna()

# model_ols = sm.OLS.from_formula('c ~ y + c1',df)
model_ols = IV2SLS.from_formula('c ~ 1 + y + c1',df)
ols1 = model_ols.fit(cov_type='unadjusted')
ols2 = model_ols.fit()  # default cov_type='robust'
print(ols2.summary)

model_iv = IV2SLS.from_formula('c ~ 1 + y + [c1 ~ y1 + y2]',df)
iv1 = model_iv.fit(cov_type='unadjusted') # try HCV-robust
print(iv1.summary)
iv2 = model_iv.fit() # default cov_type='robust'
print(iv2.summary)
# sargan test of overidentifying restrictions
print(iv2.sargan)

model_gmm = IVGMM.from_formula('c ~ 1 + y + [c1 ~ y1 + y2]',df)
gmm1 = model_gmm.fit() # 2-step, weight_type='robust'
print(gmm1.summary)
gmm2 = model_gmm.fit(iter_limit=50)
print(gmm2.summary)
# initial_weight=np.eye(3) may be used to set the weight matrix in the first iteration
model_gmmcue = IVGMMCUE.from_formula('c ~ 1 + y + [c1 ~ y1 + y2]',df)
gmm3 = model_gmmcue.fit() # weight_type='robust'
print(gmm3.summary)

# J-test of overidentifying restrictions: E(moment conditions) = 0
print(gmm3.j_stat)

# compare model results based on linearmodels
from linearmodels.iv.results import compare
res = {'OLS':ols2,'IV':iv2,'GMM-2steps':gmm1,'GMM-Iterative':gmm2,'GMM-CUE':gmm3}
print(compare(res))

# alternatively, using 
# from statsmodels.sandbox.regression.gmm import LinearIVGMM
# define xvar and ivar, then call (from_formula can not be used)
# model = LinearIVGMM(c, xvar, ivar).fit()
# see also IV2SLS, IVGMM
from statsmodels.sandbox.regression.gmm import LinearIVGMM
df = sm.add_constant(df)
yvar = df['c']
xvar = df[['const','y','c1']]
zvar = df[['const','y','y1','y2']]
model = LinearIVGMM(yvar,xvar,zvar)
linear_gmm1 = model.fit()
print(linear_gmm1.summary())  # gmm2
linear_gmm2 = model.fit(maxiter=50) # gmm2
print(linear_gmm2.summary())
linear_gmm3 = model.fit(maxiter='cue')
print(linear_gmm3.summary())
